subsys/bluetooth/mesh/foundation.h

```diff
@@ -153,18 +153,49 @@ void bt_mesh_model_reset(void);
 void bt_mesh_attention(struct bt_mesh_model *model, uint8_t time);
 
 #include <zephyr/sys/byteorder.h>
+#define KEY_UNUSED 0xffff
 
-static inline void key_idx_pack(struct net_buf_simple *buf,
-				uint16_t idx1, uint16_t idx2)
+static inline void key_idx_pack_pair(struct net_buf_simple *buf, uint16_t idx1, uint16_t idx2)
 {
 	net_buf_simple_add_le16(buf, idx1 | ((idx2 & 0x00f) << 12));
 	net_buf_simple_add_u8(buf, idx2 >> 4);
 }
 
-static inline void key_idx_unpack(struct net_buf_simple *buf,
-				  uint16_t *idx1, uint16_t *idx2)
+static inline size_t key_idx_calc_buf_len(size_t cnt)
 {
-	*idx1 = sys_get_le16(&buf->data[0]) & 0xfff;
-	*idx2 = sys_get_le16(&buf->data[1]) >> 4;
-	net_buf_simple_pull(buf, 3);
+	return 3 * ((cnt - (cnt % 2)) / 2) + (cnt % 2) * 2;
+}
+
+static inline int key_idx_pack_list(struct net_buf_simple *buf, uint16_t *arr, size_t cnt)
+{
+	int valid_key_cnt = 0;
+	uint16_t *idx = NULL;
+
+	for (int i = 0; i < cnt; i++) {
+		if (arr[i] != KEY_UNUSED) {
+			valid_key_cnt++;
+		}
+	}
+
+	if (net_buf_simple_tailroom(buf) < key_idx_calc_buf_len(valid_key_cnt)) {
```

Comment:
This shouldn't be needed because config server allocates big enough buffer: https://github.com/zephyrproject-rtos/zephyr/blob/bb74b311fe8b24afc2c6f7e5eb373fdc9ff02f6a/subsys/bluetooth/mesh/cfg_srv.c#L376-L377
If the calculation is wrong and buffer is smaller than needed, the `net_buf_simple_add_le16` will assert, which is fine since the assumption is that the buffer should always be big enough.